src/token/RestrictionManager.sol

```diff
@@ -84,33 +106,64 @@ contract RestrictionManager is Auth, IRestrictionManager, IERC20Callback {
         return SUCCESS_MESSAGE;
     }
 
-    // --- Handling freezes ---
+    // --- Incoming message handling ---
+    /// @inheritdoc IHook
+    function updateRestriction(address token, bytes memory update) external auth {
+        RestrictionUpdate updateId = RestrictionUpdate(update.toUint8(0));
+
+        if (updateId == RestrictionUpdate.UpdateMember) updateMember(token, update.toAddress(1), update.toUint64(33));
+        else if (updateId == RestrictionUpdate.Freeze) freeze(token, update.toAddress(1));
+        else if (updateId == RestrictionUpdate.Unfreeze) unfreeze(token, update.toAddress(1));
+        else revert("RestrictionManager/invalid-update");
+    }
+
     /// @inheritdoc IRestrictionManager
-    function freeze(address user) public auth {
+    function freeze(address token, address user) public auth {
         require(user != address(0), "RestrictionManager/cannot-freeze-zero-address");
-        restrictions[user].frozen = true;
-        emit Freeze(user);
+        require(!root.endorsed(user), "RestrictionManager/endorsed-user-cannot-be-frozen");
+
+        uint128 hookData = uint128(ITrancheToken(token).hookDataOf(user));
+        ITrancheToken(token).setHookData(user, bytes16(hookData.setBit(FREEZE_BIT, true)));
+
+        emit Freeze(token, user);
+    }
+
+    /// @inheritdoc IRestrictionManager
+    function unfreeze(address token, address user) public auth {
+        uint128 hookData = uint128(ITrancheToken(token).hookDataOf(user));
+        ITrancheToken(token).setHookData(user, bytes16(hookData.setBit(FREEZE_BIT, false)));
+
+        emit Unfreeze(token, user);
     }
 
     /// @inheritdoc IRestrictionManager
-    function unfreeze(address user) public auth {
-        restrictions[user].frozen = false;
-        emit Unfreeze(user);
+    function isFrozen(address token, address user) public view returns (bool) {
+        return uint128(ITrancheToken(token).hookDataOf(user)).getBit(FREEZE_BIT);
     }
 
     // --- Managing members ---
     /// @inheritdoc IRestrictionManager
-    function updateMember(address user, uint64 validUntil) public auth {
+    function updateMember(address token, address user, uint64 validUntil) public auth {
         require(block.timestamp <= validUntil, "RestrictionManager/invalid-valid-until");
-        restrictions[user].validUntil = validUntil;
+        require(!root.endorsed(user), "RestrictionManager/endorsed-user-cannot-be-updated");
+        _updateMember(token, user, validUntil);
+    }
+
+    function _updateMember(address token, address user, uint64 validUntil) internal {
```

Comment:
What's the benefit of this internal method? Why not inline the body of this function in `updateMember`?

Reply:
Good catch, this was required in a previous draft but not anymore, will change